### PR TITLE
mpath plugin: Use libdmmp API.

### PR DESCRIPTION
**Pull request by cathay4t**
* Use libdmmp for these functions of mpath plugin:
    * `bd_mpath_flush_mpaths()`
    * `bd_mpath_reconfig()`
    * `bd_mpath_is_mpath_member()`
    * `bd_mpath_get_mpath_members()`

### CODE REVIEW

File: src/plugins/mpath.c [hunk 1]
```diff
@@ -15,18 +15,57 @@
...
+        ctx = dmmp_context_new (); \
+        if (ctx != NULL) \
+          { \
+            memset(err_msg_buff, 0, _ERR_MSG_BUFF_SIZE); \
+            dmmp_context_userdata_set (ctx, err_msg_buff); \
+            dmmp_context_log_func_set (ctx, _log_func); \
+          } \
+    } while(0)
+
+#define _UNUSED(x) (void)(x)
```
Comment (vpodzime) [inline src/plugins/mpath.c]: Why do all these things start with an underscore?

File: src/plugins/mpath.c [hunk 2]
```diff
@@ -69,24 +106,15 @@ static UtilDep deps[DEPS_LAST] = {
...
-        g_warning("Cannot load the mpath plugin");
+    gboolean ret = FALSE;
 
+    /* mpathconf doesn't report its version */
+    ret = bd_utils_check_util_version ("mpathconf", NULL, NULL, NULL, &error);
+    if (!ret && error)
+      {
+        g_warning ("Cannot load the mpath plugin: %s", error->message);
+        g_clear_error (&error);
+      }
```
Comment (vpodzime) [inline src/plugins/mpath.c]: This needs to preserve the ``avail_deps`` atomic manipulation in place.

File: src/plugins/mpath.c [hunk 3]
```diff
@@ -110,7 +138,7 @@ gboolean bd_mpath_init () {
  *
  */
 void bd_mpath_close () {
-    /* nothing to do here */
+      /* nothing to do here */
```
Comment (vpodzime) [inline src/plugins/mpath.c]: Why this change?

File: src/plugins/mpath.c [hunk 4]
```diff
@@ -156,360 +184,295 @@ gboolean bd_mpath_is_tech_avail (BDMpathTech tech, guint64 mode, GError **error)
  * Tech category: %BD_MPATH_TECH_BASE-%BD_MPATH_TECH_MODE_MODIFY
  */
 gboolean bd_mpath_flush_mpaths (GError **error) {
-    const gchar *argv[3] = {"multipath", "-F", NULL};
-    gboolean success = FALSE;
-    gchar *output = NULL;
+    struct dmmp_mpath **mps = NULL;
+    struct dmmp_context *ctx = NULL;
+    uint32_t mp_count = 0;
+    uint32_t i = 0;
```
Comment (vpodzime) [inline src/plugins/mpath.c]: Please use the *GLib* types here.

File: src/plugins/mpath.c [hunk 5]
```diff
@@ -156,360 +184,295 @@ gboolean bd_mpath_is_tech_avail (BDMpathTech tech, guint64 mode, GError **error)
...
+    gboolean rc = FALSE;
+    const char *name = NULL;
+    int dmmp_rc = DMMP_OK;
+    char err_msg_buff[_ERR_MSG_BUFF_SIZE];
 
-    if (!check_deps (&avail_deps, DEPS_MPATH_MASK, deps, DEPS_LAST, &deps_check_lock, error))
-        return FALSE;
+    assert (error != NULL);
```
Comment (vpodzime) [inline src/plugins/mpath.c]: Please don't use ``assert()`` in a library.

Comment (vpodzime) [reply src/plugins/mpath.c]: We need some error to be set and ``FALSE`` be returned instead.

File: src/plugins/mpath.c [hunk 6]
```diff
@@ -156,360 +184,295 @@ gboolean bd_mpath_is_tech_avail (BDMpathTech tech, guint64 mode, GError **error)
...
-        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_INVAL,
-                     "Failed to determine device name for '%s'",
-                     major_minor);
-        g_free (link);
-        return NULL;
+    _dmmp_ctx_setup (ctx, err_msg_buff);
+    if (ctx == NULL) {
+        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_NO_MEMORY,
+                     "No enough memory to create dmmp context");
+        goto out;
```
Comment (vpodzime) [inline src/plugins/mpath.c]: Please avoid the ``goto`` statement. I know it make things more complicated, but we have always been sacrificing that for avoiding ``goto``s in this library and this doesn't look like any special case that should end such efforts.

File: src/plugins/mpath.c [hunk 7]
```diff
@@ -156,360 +184,295 @@ gboolean bd_mpath_is_tech_avail (BDMpathTech tech, guint64 mode, GError **error)
...
 
-    if (dm_task_get_info (task, &info) == 0) {
-        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_DM_ERROR,
-                     "Failed to get task info");
-        dm_task_destroy (task);
-        return FALSE;
+out:
+    if (dmmp_rc != DMMP_OK) {
+        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_FLUSH,
+                     "%d: %s", dmmp_rc, _dmmp_last_err_msg(ctx));
```
Comment (vpodzime) [inline src/plugins/mpath.c]: This should provide some more descriptive error messages based on what (step) exactly failed.

File: src/plugins/mpath.c [hunk 8]
```diff
@@ -156,360 +184,295 @@ gboolean bd_mpath_is_tech_avail (BDMpathTech tech, guint64 mode, GError **error)
...
-    if (geteuid () != 0) {
-        g_set_error (error, BD_MPATH_ERROR, BD_MPATH_ERROR_NOT_ROOT,
-                     "Not running as root, cannot query DM maps");
-        bd_utils_report_finished (progress_id, (*error)->message);
-        return NULL;
+    GList *path_blk_names = NULL;
+    gchar *dup = NULL;
+    guint len = 0;
+    guint m = 0;
+    GList *cur_node = NULL;
```
Comment (vpodzime) [inline src/plugins/mpath.c]: Please use a ``GPtrArray`` here. And note that ``g_ptr_array_free (arr, FALSE)`` can be used to free the container while preserving the content which means the content doesn't have to be copied over elsewhere.

File: src/plugins/mpath.h [hunk 9]
```diff
@@ -25,7 +28,6 @@ typedef enum {
     BD_MPATH_TECH_MODE_MODIFY = 1 << 1,
 } BDMpathTechMode;
 
-
```
Comment (vpodzime) [inline src/plugins/mpath.h]: no need for this change

**Comment by cathay4t** [comment]
Fedora 27 and later has the `libdmmp` function required by this PR.

File: src/plugins/mpath.c (hunk 1 above)
Comment (cathay4t) [reply src/plugins/mpath.c]: Just don't want to override existing macros (from other header) by accident.
Just a habit of keep internal functions start with an underscore.

File: src/plugins/mpath.c (hunk 3 above)
Comment (cathay4t) [reply src/plugins/mpath.c]: Just a trivial miss type. Will remove it.

File: src/plugins/mpath.c (hunk 4 above)
Comment (cathay4t) [reply src/plugins/mpath.c]: These are for `libdmmp` function only. Why we need `GLib`ize them?

File: src/plugins/mpath.c (hunk 5 above)
Comment (cathay4t) [reply src/plugins/mpath.c]: Error cannot be set due to NULL pointer. Will just return `FALSE`.

File: src/plugins/mpath.c (hunk 6 above)
Comment (cathay4t) [reply src/plugins/mpath.c]: Sure. Will do.

File: src/plugins/mpath.c (hunk 7 above)
Comment (cathay4t) [reply src/plugins/mpath.c]: Will fix this by removing `goto` statement.

File: src/plugins/mpath.c (hunk 6 above)
Comment (cathay4t) [reply src/plugins/mpath.c]: Just FYI: https://peerj.com/preprints/826v1/ 
```
We conclude that developers limit themselves to using 
goto appropriately in most cases, and not in an unrestricted 
manner like Dijkstra feared, thus suggesting that goto does 
not appear to be harmful in practice.
```

File: src/plugins/mpath.c (hunk 8 above)
Comment (cathay4t) [reply src/plugins/mpath.c]: The memory holding the path device string is managed/owned by `struct dmmp_mpath` where we cannot only partly free(free all memory except the path device string). Hence `strdup()` is required.

**Comment by cathay4t** [comment]
V2 patch uploaded.

Changes:
 * Remove `goto` statement.
 * Remove requirement of `multipath` tool. Always return `TRUE` for `BD_MPATH_TECH_BASE`. The libdmmp only require a running `multipathd`.
 * Fix issue found by vpodzime

Patch was tested by https://gist.github.com/cathay4t/7788a8dcaa3ba59bea1a862702cf07cf

**Comment by vpodzime** [comment]
This PR is abandoned.
